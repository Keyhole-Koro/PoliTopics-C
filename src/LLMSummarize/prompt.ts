export const instruction = `以下の会話内容をもとに、次の形式で要約データを構成してください:

各セクションとフィールドの**使用目的**を理解した上で、忠実に出力してください。  
この構造は、人間が読むレポートとしてだけでなく、システムが発言と要約を対応付けて処理できるように設計されています。

---

1. **基本情報 (Metadata)**  
 会議に関するメタデータを記載してください:タイトル、開催日、開催機関、カテゴリなど。  

2. **全体の要約 (Summary)**  
 会議全体の要点や結論を簡潔にまとめてください。  
 - \`based_on_orders\`: この要約がどの dialogs.order をまとめているかを配列で示す (例:[1,2,3])。  
 - \`summary\`: 会話全体の要点。  
 - \`figure\`: 補足的な図解 (任意)。  

3. **やさしい要約 (SoftSummary)**  
 政治や専門用語に馴染みのない読者向けに、背景や文脈も含めてわかりやすく丁寧に説明してください。  
 - \`based_on_orders\`: 対象となる dialogs.order の配列。  
 - \`summary\`: わかりやすく言い換えた文章。  

4. **中間要約 (MiddleSummary)**  
 議論の重要な転換点や話題ごとのまとまりを要約してください。構成順に並べてください。  
 - \`based_on_orders\`: どの dialogs.order に基づくか。  
 - \`summary\`: 中間的なまとめ。  
 - \`figure\`: 図や補足 (任意)。  

5. **発言ごとの要約 (Dialogs)**  
 各発言について、以下の情報を含めて記述してください:  
 - \`order\`: 発言番号。  
 - \`speaker\`: 発言者名。  
 - \`speaker_group\`: 所属。  
 - \`speaker_position\`: 役職。  
 - \`speaker_role\`: 役割 (質問者、答弁者など)。  
 - \`summary\`: 発言の主旨を簡潔に要約。  
 - \`soft_summary\`: 一般読者にも伝わるように、やさしく丁寧に言い換えた文章。  
 - \`response_to\`: この発言がどの発言に反応しているか (例:質問、賛同、反論など)。  

6. **参加者情報 (Participants)**  
 主な話者ごとに、名前・役職・発言内容の要旨をまとめてください。  

7. **用語の解説 (Terms)**  
 専門的または一般にはわかりにくい用語について、簡潔で明確な定義を記述してください。文脈に即した説明が望ましいです。  

8. **キーワード抽出 (Keywords)**  
 議論の焦点となる用語やトピックを抽出し、重要度 (high / medium / low)を分類してください。  

---
`;

export const output_format = `### 出力フォーマット

以下の形式に従ってJSONデータを作成してください:

{
  "id": "文字列 (議事録ID)",
  "title": "会議のタイトル",
  "date": "開催日 (YYYY-MM-DD)",
  "imageKind": "画像分類 (例: graph, diagram, etc.)",
  "session": 数字 (例: 208),
  "nameOfHouse": "衆議院または参議院",
  "nameOfMeeting": "会議名 (例: 国土交通委員会)",
  "category": "カテゴリ (例: 環境, 教育, etc.)",
  "description": "この会議についての説明",

  "summary": {
    "based_on_orders": [1,2,3],
    "summary": "会話全体の要約をここに記載",
    "figure": "Markdown形式で図や補足を記載 (任意)"
  },
  "soft_summary": {
    "based_on_orders": [1,2,3],
    "summary": "政治に詳しくない人でも分かるように、丁寧でわかりやすく説明した要約"
  },
  "middle_summary": [
    {
      "based_on_orders": [4,5],
      "summary": "中間要約1",
      "figure": "Markdown形式 (任意)"
    }
  ],
  "dialogs": [
    {
      "order": 1,
      "speaker": "発言者名",
      "speaker_group": "所属",
      "speaker_position": "役職",
      "speaker_role": "役割",
      "summary": "発言内容の要約",
      "soft_summary": "発言をやさしく丁寧に言い換えた内容",
      "response_to": [
        {
          "order": 0,
          "reaction": "agree | disagree | neutral | question | answer"
        }
      ]
    }
  ],
  "participants": [
    {
      "name": "話者名",
      "summary": "この人の発言要旨"
    }
  ],
  "terms": [
    {
      "term": "専門用語",
      "definition": "その説明"
    }
  ],
  "keywords": [
    {
      "keyword": "キーワード",
      "priority": "high | medium | low"
    }
  ]
}
`;

export const compose_prompt = (content: string): string => {
  return `${instruction}\n${output_format}\n${content}`;
};
