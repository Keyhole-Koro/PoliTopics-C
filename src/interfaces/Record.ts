export interface Speech {
    speechOrder: number;
    speaker: string;
    speakerYomi: string;
    speakerGroup: string;
    speakerPosition: string;
    speakerRole: string;
    speech: string;
}

export interface MapIssue {
    imageKind: string;
    session: number;
    nameOfHouse: string;
    nameOfMeeting: string;
    date: string;
    speeches: Speech[];
}

// Type guards for runtime type checking
export const isSpeech = (obj: any): obj is Speech => {
    return obj &&
        typeof obj.speechOrder === 'number' &&
        typeof obj.speaker === 'string' &&
        typeof obj.speakerYomi === 'string' &&
        typeof obj.speakerGroup === 'string' &&
        typeof obj.speakerPosition === 'string' &&
        typeof obj.speakerRole === 'string' &&
        typeof obj.speech === 'string';
};
