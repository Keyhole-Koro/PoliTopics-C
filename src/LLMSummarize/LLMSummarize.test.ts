import 'dotenv/config';
import fs from 'fs';
import path from 'path';
import SpeechFormatter from '@NationalDietRecord/recordFormat';
import { RawData } from '@NationalDietRecord/RawRecord';
import { MapIssue } from '@interfaces/Record';

import geminiAPI from './gemini';
import { compose_prompt } from './prompt';

describe('Gemini API Handler', () => {
    const GEMINI_API_KEY = process.env.GEMINI_API_KEY || '';
    let formattedData: Record<string, MapIssue>;

    beforeAll(() => {
        // Load and format the fetchedRecord.json file
        const filePath: string = path.resolve(__dirname, 'fetchedRecord.json');
        const rawData: RawData = JSON.parse(fs.readFileSync(filePath, 'utf-8'));

        const formatter: SpeechFormatter = new SpeechFormatter();
        formattedData = formatter.mapRecords(rawData);
    });

    it('should generate summaries for formatted records', async () => {
        for (const [issueId, issue] of Object.entries(formattedData)) {
            if (issueId === '121714024X01920250514') continue; // Skip this specific issue
            const issueText = JSON.stringify(issue, null, 2);
            const result = await geminiAPI(GEMINI_API_KEY, compose_prompt(issueText));
            const resultJson = JSON.parse(result);
    
            console.log(`Generated Summary for Issue ${issueId}:`, resultJson, typeof resultJson);
    
            // Validate resultJson matches the Article structure
            expect(resultJson).toBeDefined();
            expect(typeof resultJson).toBe('object');
            /*
    
            // Check if resultJson matches the Article structure
            expect(resultJson).toHaveProperty('title');
            console.log(`Title: ${resultJson.title}, Type: ${typeof resultJson.title}`);
            expect(typeof resultJson.title).toBe('string');
    
            expect(resultJson).toHaveProperty('summary');
            expect(typeof resultJson.summary).toBe('string');
    
            expect(resultJson).toHaveProperty('content');
            expect(typeof resultJson.content).toBe('string');

    
            // Optional fields
            if (resultJson.author) {
                expect(typeof resultJson.author).toBe('string');
            }
            if (resultJson.date) {
                expect(typeof resultJson.date).toBe('string');
            }
            */
        }
    }, 50000);
});
