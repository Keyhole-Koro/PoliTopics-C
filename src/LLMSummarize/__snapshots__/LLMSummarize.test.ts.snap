// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gemini API Handler should generate summaries for formatted records: Gemini response for issue 121714024X01820250509 1`] = `
{
  "category": "政治",
  "date": "2025-05-09",
  "description": "参議院議院運営委員会における会議録。重要経済安保情報に関する法律案の採決、規則案・規程案の委員会審査省略要求の取扱い、国会職員の適性評価に関する件などが議題となった。",
  "dialogs": [
    {
      "order": 1,
      "response_to": [],
      "soft_summary": "会議の日時や出席者、そして会議で話し合われたことの一覧です。",
      "speaker": "会議録情報",
      "speaker_group": null,
      "speaker_position": null,
      "speaker_role": null,
      "summary": "会議の開催日時、出席者リスト、議題のリスト。",
    },
    {
      "order": 2,
      "response_to": [],
      "soft_summary": "会議が始まり、最初に話し合うこととして、国の重要な経済や安全保障に関する秘密を守るための法律案が決められました。その法律案について、詳しい説明を聞くことになりました。",
      "speaker": "牧野たかお",
      "speaker_group": "自由民主党",
      "speaker_position": "委員長",
      "speaker_role": "議長",
      "summary": "会議の開会宣言と、重要経済安保情報に関する法律案の議題決定、趣旨説明の依頼。",
    },
    {
      "order": 3,
      "response_to": [
        {
          "id": 2,
          "reaction": "answer",
        },
      ],
      "soft_summary": "国の重要な経済や安全保障に関する秘密を守るための法律案について、詳しい説明がありました。この法律は、5月16日から施行されます。",
      "speaker": "浜田靖一",
      "speaker_group": "自由民主党・無所属の会",
      "speaker_position": null,
      "speaker_role": "衆議院議員",
      "summary": "重要経済安保情報に関する法律案の趣旨説明。施行日は令和7年5月16日。",
    },
    {
      "order": 4,
      "response_to": [
        {
          "id": 3,
          "reaction": "neutral",
        },
      ],
      "soft_summary": "法律案について、何か意見のある人は発言するように促しました。",
      "speaker": "牧野たかお",
      "speaker_group": "自由民主党",
      "speaker_position": "委員長",
      "speaker_role": "議長",
      "summary": "法律案に関する意見聴取の呼びかけ。",
    },
    {
      "order": 5,
      "response_to": [
        {
          "id": 4,
          "reaction": "disagree",
        },
      ],
      "soft_summary": "国の重要な秘密を守るための法律案に反対する意見が述べられました。この法律案によって、国民の権利が制限されたり、政治家たちが自由に発言できなくなったりするのではないかと懸念されています。",
      "speaker": "岩渕友",
      "speaker_group": "日本共産党",
      "speaker_position": null,
      "speaker_role": "委員",
      "summary": "重要経済安保情報に関する法律案への反対意見表明。国民主権と基本的人権を侵害する、国会の情報監視機能を弱体化させるものと批判。",
    },
    {
      "order": 6,
      "response_to": [
        {
          "id": 5,
          "reaction": "neutral",
        },
      ],
      "soft_summary": "法律案について投票が行われ、賛成多数で可決されました。",
      "speaker": "牧野たかお",
      "speaker_group": "自由民主党",
      "speaker_position": "委員長",
      "speaker_role": "議長",
      "summary": "法律案の採決。賛成多数で可決。",
    },
    {
      "order": 7,
      "response_to": [
        {
          "id": 6,
          "reaction": "neutral",
        },
      ],
      "soft_summary": "法律案の詳しい内容をまとめた報告書の作成は、委員長にお任せすることに決まりました。",
      "speaker": "牧野たかお",
      "speaker_group": "自由民主党",
      "speaker_position": "委員長",
      "speaker_role": "議長",
      "summary": "審査報告書作成の委員長への一任決定。",
    },
    {
      "order": 8,
      "response_to": [
        {
          "id": 7,
          "reaction": "neutral",
        },
      ],
      "soft_summary": "会議のやり方に関するルール改正案と、会議の運営に関するルール改正案をまとめて話し合うことになり、事務局から説明を聞きました。",
      "speaker": "牧野たかお",
      "speaker_group": "自由民主党",
      "speaker_position": "委員長",
      "speaker_role": "議長",
      "summary": "規則案と規程案の議題決定、事務総長への報告依頼。",
    },
    {
      "order": 9,
      "response_to": [
        {
          "id": 8,
          "reaction": "answer",
        },
      ],
      "soft_summary": "会議のやり方に関するルール改正案と、会議の運営に関するルール改正案について、これらを簡単に決めてしまいたいという要望がありました。",
      "speaker": "小林史武",
      "speaker_group": null,
      "speaker_position": "参議院事務総長",
      "speaker_role": "事務局",
      "summary": "規則案と規程案に関する報告。委員会審査省略要求あり。",
    },
    {
      "order": 10,
      "response_to": [
        {
          "id": 9,
          "reaction": "neutral",
        },
      ],
      "soft_summary": "ルール改正案について投票が行われ、賛成多数で簡単に決めることになりました。",
      "speaker": "牧野たかお",
      "speaker_group": "自由民主党",
      "speaker_position": "委員長",
      "speaker_role": "議長",
      "summary": "規則案と規程案の委員会審査省略に関する採決。賛成多数で決定。",
    },
    {
      "order": 11,
      "response_to": [
        {
          "id": 10,
          "reaction": "neutral",
        },
      ],
      "soft_summary": "国会議員を助ける職員の評価方法について話し合うことになり、事務局から説明を聞きました。",
      "speaker": "牧野たかお",
      "speaker_group": "自由民主党",
      "speaker_position": "委員長",
      "speaker_role": "議長",
      "summary": "国会職員の適性評価に関する件の議題決定、事務総長への説明依頼。",
    },
    {
      "order": 12,
      "response_to": [
        {
          "id": 11,
          "reaction": "answer",
        },
      ],
      "soft_summary": "国会議員を助ける職員の評価方法について、国の重要な経済や安全保障に関する秘密を守るための規定を追加することになりました。",
      "speaker": "小林史武",
      "speaker_group": null,
      "speaker_position": "参議院事務総長",
      "speaker_role": "事務局",
      "summary": "国会職員の適性評価に関する説明。重要経済安保情報に関する規定を追加。",
    },
  ],
  "imageKind": "会議録",
  "issueID": "217-20250509",
  "keywords": [
    {
      "keyword": "重要経済安保情報",
      "priority": "high",
    },
    {
      "keyword": "国会法改正",
      "priority": "high",
    },
    {
      "keyword": "秘密保護",
      "priority": "high",
    },
    {
      "keyword": "適性評価",
      "priority": "medium",
    },
    {
      "keyword": "委員会審査省略",
      "priority": "medium",
    },
    {
      "keyword": "日本共産党",
      "priority": "medium",
    },
  ],
  "middle_summary": [
    {
      "figure": null,
      "order": 1,
      "summary": "重要経済安保情報に関する法律案の趣旨説明と採決。日本共産党から反対意見が出された。",
    },
    {
      "figure": null,
      "order": 2,
      "summary": "規則案と規程案の委員会審査省略に関する件。委員長提案で可決。",
    },
    {
      "figure": null,
      "order": 3,
      "summary": "国会職員の適性評価に関する件の一部改正。事務総長による説明の後、決定。",
    },
  ],
  "nameOfHouse": "参議院",
  "nameOfMeeting": "議院運営委員会",
  "participants": [
    {
      "name": "牧野たかお",
      "summary": "委員長として会議を進行し、議題決定、意見聴取、採決などを担当。",
    },
    {
      "name": "浜田靖一",
      "summary": "重要経済安保情報に関する法律案の趣旨説明を行った。",
    },
    {
      "name": "岩渕友",
      "summary": "重要経済安保情報に関する法律案に反対し、その理由を詳細に説明した。",
    },
    {
      "name": "小林史武",
      "summary": "事務総長として、規則案・規程案、国会職員の適性評価に関する説明を行った。",
    },
  ],
  "session": 217,
  "soft_summary": {
    "id": 1,
    "summary": "国の重要な秘密を守るための法律案が参議院で承認されました。この法律案は、経済に関する重要な秘密をしっかり守るためのものですが、反対する意見もありました。反対する人は、この法律によって、政治家たちが自由に発言できなくなったり、重要な情報が国民に届かなくなったりするのではないかと心配しています。他にも、会議のやり方に関するルール改正や、国会議員を助ける職員の評価方法に関する話もされました。",
  },
  "summary": {
    "figure": null,
    "id": 1,
    "summary": "参議院議院運営委員会において、重要経済安保情報の保護に関する法律案が可決された。また、規則案・規程案の委員会審査省略、国会職員の適性評価に関する事項も審議され、決定された。日本共産党の岩渕友委員は、これらの改正案が国会の情報監視機能を弱体化させ、議員の発言・質問の自由を制限すると強く批判した。",
  },
  "terms": [
    {
      "definition": "経済安全保障上重要な情報。国家機密に準ずる扱いとなる。",
      "term": "重要経済安保情報",
    },
    {
      "definition": "個人情報の把握による、特定の職務への適任性の評価。",
      "term": "適性評価",
    },
    {
      "definition": "委員会での審議を省略し、迅速に決定を行うこと。",
      "term": "委員会審査省略",
    },
  ],
  "title": "参議院議院運営委員会",
}
`;
